R/glex.R

```diff
@@ -115,66 +111,246 @@ glex.xgb.Booster <- function(object, x, max_interaction = NULL, ...) {
 
-  # Function to get all subsets of set
-  subsets <- function(x) {
-    if (length(x) == 1) {
-      list(integer(0), x)
+  # Calculate components
+  res <- calc_components(trees, x, max_interaction, features, probFunction)
+  res$intercept <- res$intercept + 0.5
+  
+  # Return components
+  res
+} 
+
+#' @rdname glex
+#' @export
+#' @useDynLib glex, .registration = TRUE
+#' @import Rcpp
+#' @import data.table
+#' @import foreach
+#' @importFrom stats predict
+#' @importFrom utils combn
+#'
+#' @examples
+#' # ranger -----
+#' if (requireNamespace("ranger", quietly = TRUE)) {
+#' library(ranger)
+#' x <- as.matrix(mtcars[, -1])
+#' y <- mtcars$mpg
+#' rf <- ranger(x = x[1:26, ], y = y[1:26],
+#'              num.trees = 5, max.depth = 3, 
+#'              node.stats = TRUE)
+#' glex(rf, x[27:32, ])
+#'
+#' \dontrun{
+#' # Parallel execution
+#' doParallel::registerDoParallel()
+#' glex(rf, x[27:32, ])
+#' }
+#' }
+glex.ranger <- function(object, x, max_interaction = NULL, features = NULL, probFunction = NULL, ...) {
+
+  # To avoid data.table check issues
+  terminal <- NULL
+  splitvarName <- NULL
+  splitStat <- NULL
+  prediction <- NULL
+  splitvarID <- NULL
+  tree <- NULL
+  
+  if (!requireNamespace("ranger", quietly = TRUE)) {
+    stop("ranger needs to be installed: install.packages(\"ranger\")")
+  }
+  
+  if (is.null(object$forest$num.samples.nodes)) {
+    stop("ranger needs to be called with node.stats=TRUE for glex.")
+  }
+  
+  # If max_interaction is not specified, we set it to the max.depth param of the ranger model.
+  # If max.depth is not defined in ranger, we assume 6 as in xgboost.
+  rf_max_depth <- ifelse((is.null(object$max.depth) || object$max.depth == 0), 6L, object$max.depth)
+  
+  if (is.null(max_interaction)) {
+    max_interaction <- rf_max_depth
+  }
+  
+  checkmate::assert_int(max_interaction, lower = 1, upper = Inf)
+  
+  # Convert model into xgboost format
+  trees <- rbindlist(lapply(seq_len(object$num.trees), function(i) {
+    as.data.table(ranger::treeInfo(object, tree = i))[, tree := i-1]
+  }))
+  trees[terminal == TRUE, splitvarName := "Leaf"]
+  trees[terminal == TRUE, splitStat := prediction]
+  trees[, splitvarID := NULL]
+  trees[, terminal := NULL]
+  trees[, prediction := NULL]
+  colnames(trees) <- c("Node", "Yes", "No", "Feature", "Split", "Cover", "Quality", "Tree")
+  
+  # Calculate components
+  res <- calc_components(trees, x, max_interaction, features, probFunction)
+  
+  # Divide everything by the number of trees
+  res$shap <- res$shap / object$num.trees
+  res$m <- res$m / object$num.trees
+  res$intercept <- res$intercept / object$num.trees
+  
+  # Return components
+  res
+} 
+
+# Function to get all subsets of set
+subsets <- function(x) {
+  if (length(x) == 1) {
+    list(integer(0), x)
+  } else {
+    do.call(c, lapply(0:length(x), combn, x = x, simplify = FALSE))
+  }
+}
+
+tree_fun_old <- function(tree, trees, x, all_S) {
```

Comment:
Maybe rename this to something more informative? Functions with `_old` suffixes tend to just make me assume they're obsolete or deprecated, which doesn't appear to be the case here.

Thanks for prying out the internal functions though 😬